File: crates/api/src/handlers/tags.rs
//! `tags` domain handlers.
//!
//! Wave-E2.4 (Olga). Five-command CRUD plus the `prompt_tags`
//! join-table helpers (`add_prompt_tag` / `remove_prompt_tag`).

use catique_application::{tags::TagsUseCase, AppError};
use catique_domain::Tag;
use catique_infrastructure::db::{pool::acquire, repositories::tags as repo};
use serde_json::json;
use tauri::State;

use crate::events;
use crate::state::AppState;

/// E2 will populate per-domain initialisation here.
pub fn register() {}

/// IPC: list every tag.
///
/// # Errors
///
/// Forwards every error from `TagsUseCase::list`.
#[tauri::command]
pub async fn list_tags(state: State<'_, AppState>) -> Result<Vec<Tag>, AppError> {
    TagsUseCase::new(&state.pool).list()
}

/// IPC: look up a tag by id.
///
/// # Errors
///
/// Forwards every error from `TagsUseCase::get`.
#[tauri::command]
pub async fn get_tag(state: State<'_, AppState>, id: String) -> Result<Tag, AppError> {
    TagsUseCase::new(&state.pool).get(&id)
}

/// IPC: create a tag.
///
/// # Errors
///
/// Forwards every error from `TagsUseCase::create`.
#[tauri::command]
pub async fn create_tag(
    state: State<'_, AppState>,
    name: String,
    color: Option<String>,
) -> Result<Tag, AppError> {
    let tag = TagsUseCase::new(&state.pool).create(name, color)?;
    events::emit(&state, events::TAG_CREATED, json!({ "id": tag.id }));
    Ok(tag)
}

/// IPC: partial-update a tag.
///
/// # Errors
///
/// Forwards every error from `TagsUseCase::update`.
#[tauri::command]
pub async fn update_tag(
    state: State<'_, AppState>,
    id: String,
    name: Option<String>,
    color: Option<Option<String>>,
) -> Result<Tag, AppError> {
    let tag = TagsUseCase::new(&state.pool).update(id, name, color)?;
    events::emit(&state, events::TAG_UPDATED, json!({ "id": tag.id }));
    Ok(tag)
}

/// IPC: delete a tag.
///
/// # Errors
///
/// Forwards every error from `TagsUseCase::delete`.
#[tauri::command]
pub async fn delete_tag(state: State<'_, AppState>, id: String) -> Result<(), AppError> {
    TagsUseCase::new(&state.pool).delete(&id)?;
    events::emit(&state, events::TAG_DELETED, json!({ "id": id }));
    Ok(())
}

// ---------------------------------------------------------------------
// Join-table helpers — prompt_tags.
// ---------------------------------------------------------------------

/// Attach a tag to a prompt.
///
/// # Errors
///
/// `AppError::TransactionRolledBack` on FK violation.
#[tauri::command]
pub async fn add_prompt_tag(
    state: State<'_, AppState>,
    prompt_id: String,
    tag_id: String,
) -> Result<(), AppError> {
    let conn = acquire(&state.pool).map_err(map_db)?;
    repo::add_prompt_tag(&conn, &prompt_id, &tag_id).map_err(map_db)
}

/// Detach a tag from a prompt.
///
/// # Errors
///
/// `AppError::NotFound` if no row matched.
#[tauri::command]
pub async fn remove_prompt_tag(
    state: State<'_, AppState>,
    prompt_id: String,
    tag_id: String,
) -> Result<(), AppError> {
    let conn = acquire(&state.pool).map_err(map_db)?;
    let removed = repo::remove_prompt_tag(&conn, &prompt_id, &tag_id).map_err(map_db)?;
    if removed {
        Ok(())
    } else {
        Err(AppError::NotFound {
            entity: "prompt_tag".into(),
            id: format!("{prompt_id}|{tag_id}"),
        })
    }
}

fn map_db(err: catique_infrastructure::db::pool::DbError) -> AppError {
    use catique_infrastructure::db::pool::DbError;
    match err {
        DbError::PoolTimeout(_) | DbError::Pool(_) => AppError::DbBusy,
        DbError::Sqlite(e) => AppError::TransactionRolledBack {
            reason: e.to_string(),
        },
        DbError::Io(e) => AppError::TransactionRolledBack {
            reason: e.to_string(),
        },
    }
}

File: src/e2e/bridge/handlers/tags.ts
/**
 * Tags command dispatcher (CRUD + the prompt_tags join helpers).
 */

import type { Tag } from "@bindings/Tag";

import { emitEvent } from "../events";
import { nextId } from "../ids";
import { store } from "../store";
import { nowBig } from "../time";

interface CreateTagArgs {
  name: string;
  color?: string | null;
}

interface UpdateTagArgs {
  id: string;
  name?: string;
  color?: string | null;
}

export function handleTags(
  command: string,
  args: Record<string, unknown>,
): unknown {
  switch (command) {
    case "list_tags":
      return Array.from(store.tags.values()).sort((a, b) =>
        a.name.localeCompare(b.name),
      );
    case "get_tag": {
      const id = String(args["id"]);
      const t = store.tags.get(id);
      if (!t) {
        throw {
          kind: "notFound",
          data: { entity: "tag", id },
        };
      }
      return t;
    }
    case "create_tag": {
      const a = args as unknown as CreateTagArgs;
      const id = nextId("tag");
      const ts = nowBig();
      const tag: Tag = {
        id,
        name: a.name,
        color: a.color ?? null,
        createdAt: ts,
        updatedAt: ts,
      };
      store.tags.set(id, tag);
      emitEvent("tag:created", { id });
      return tag;
    }
    case "update_tag": {
      const a = args as unknown as UpdateTagArgs;
      const prev = store.tags.get(a.id);
      if (!prev) {
        throw {
          kind: "notFound",
          data: { entity: "tag", id: a.id },
        };
      }
      const next: Tag = {
        ...prev,
        ...(a.name !== undefined ? { name: a.name } : {}),
        ...(a.color !== undefined ? { color: a.color } : {}),
        updatedAt: nowBig(),
      };
      store.tags.set(a.id, next);
      emitEvent("tag:updated", { id: a.id });
      return next;
    }
    case "delete_tag": {
      const id = String(args["id"]);
      store.tags.delete(id);
      // Detach from every prompt to keep snapshot consistent.
      for (const [pid, tagIds] of store.promptTags) {
        store.promptTags.set(
          pid,
          tagIds.filter((x) => x !== id),
        );
      }
      emitEvent("tag:deleted", { id });
      return null;
    }
    case "add_prompt_tag": {
      const promptId = String(args["promptId"]);
      const tagId = String(args["tagId"]);
      const list = store.promptTags.get(promptId) ?? [];
      if (!list.includes(tagId)) list.push(tagId);
      store.promptTags.set(promptId, list);
      return null;
    }
    case "remove_prompt_tag": {
      const promptId = String(args["promptId"]);
      const tagId = String(args["tagId"]);
      store.promptTags.set(
        promptId,
        (store.promptTags.get(promptId) ?? []).filter((x) => x !== tagId),
      );
      return null;
    }
    case "set_tag_prompts":
      return null;
    default:
      return undefined;
  }
}
